### PR TITLE
DIRSERVER-2326 Use Bouncy Castle to generate certificates

### PR DESCRIPTION
**Pull request by kwart**
One of the possible fixes for https://issues.apache.org/jira/browse/DIRSERVER-2326

This one uses Bouncy Castle PKIX library (+860KB).
Check also another possible solution in #41.

### CODE REVIEW

**Comment by coheigea** [comment]
Looks like there are some failing tests in osgi-integ with this patch.

**Comment by kwart** [comment]
I've fixed the `core/pom.xml` in the second commit. It resolves the failures in `osgi-test` modules.

File: all/pom.xml [hunk 1]
```diff
@@ -163,11 +163,11 @@
               <promoteTransitiveDependencies>true</promoteTransitiveDependencies>
               <filters>
                 <filter>
-                  <artifact>org.bouncycastle:bcprov-jdk15on</artifact>
+                  <artifact>*:*</artifact>
```
Comment (coheigea) [inline all/pom.xml:166]: Couldn't we keep it as org.bouncycastle:*

File: service/pom.xml [hunk 2]
```diff
@@ -108,10 +108,11 @@
             <configuration>
               <filters>
                 <filter>
-                  <artifact>org.bouncycastle:bcprov-jdk15on</artifact>
+                  <artifact>*:*</artifact>
```
Comment (coheigea) [inline service/pom.xml:111]: Same here.

**Review by coheigea — CHANGES_REQUESTED** [review]

File: all/pom.xml (hunk 1 above)
Comment (kwart) [reply all/pom.xml:166]: We can keep the smaller scope, but I think we never want to include the signature files. So upgrade of a library which suddenly starts to sign its JARs wouldn't require the filter change.

@coheigea So do you think we should have the smaller scope and filter the libraries with signed classes if a need comes?

Comment (elecharny) [reply all/pom.xml:166]: Makes me think that we have some trouble with the jar we are delivering, since M24: they are not signed anymore...

Comment (coheigea) [reply all/pom.xml:166]: @kwart Yeah you're right, let's leave it as it is.

File: core/src/main/java/org/apache/directory/server/core/security/CertificateUtil.java [hunk 3]
```diff
@@ -81,94 +75,33 @@ private CertificateUtil()
...
-        info.set( X509CertInfo.KEY, new CertificateX509Key( keyPair.getPublic() ) );
-
-        // Extensions. Basically, a subjectAltName and a Basic-Constraint 
-        CertificateExtensions extensions = new CertificateExtensions();
-
-        // SubjectAltName
-        GeneralNames names = new GeneralNames();
-        names.add( new GeneralName( new DNSName( InetAddress.getLocalHost().getHostName() ) ) );
-        String ipAddress = InetAddress.getLocalHost().getHostAddress();
-        names.add( new GeneralName( new IPAddressName( ipAddress ) ) );
```
Comment (coheigea) [inline core/src/main/java/org/apache/directory/server/core/security/CertificateUtil.java:152]: I'm wondering if we should preserve the subject alt name extension?

Comment (kwart) [reply core/src/main/java/org/apache/directory/server/core/security/CertificateUtil.java:152]: Good point, I'm going to have a look.

Comment (kwart) [reply core/src/main/java/org/apache/directory/server/core/security/CertificateUtil.java:152]: Added the SAN extension in the new commit.

**Review by coheigea — APPROVED** [review]

**Comment by kwart** [comment]
@coheigea Can we get this PR merged?

**Pull request merged**